feat(dashboard): NV 5346 add search functionality to locale and timezone selects

BiswaViraj opened the pull request:
### What changed? Why was the change needed?
https://linear.app/novu/issue/NV-5346/locale-optimisation
<!-- Also include any relevant links, such as Linear tickets, Slack discussions, or design documents. -->

### Screenshots

https://github.com/user-attachments/assets/7d1422a0-9a11-4a58-ad4f-7b654de1fab5


<!-- If the changes are visual, include screenshots or screencasts. -->

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Related enterprise PR

<!-- A link to a dependent pull request  -->

### Special notes for your reviewer

<!-- Specific instructions or considerations you want to highlight for the reviewer. -->

</details>

BiswaViraj committed:
feat(dashboard): Implement edit subscriber functionality with routing and UI

BiswaViraj committed:
feat(subscriber): Add subscriber fetching functionality and UI components

BiswaViraj committed:
feat(subscriber): Implement patch subscriber functionality and update UI components

BiswaViraj committed:
feat(subscriber): Update subscriber schema to allow nullable email and add loading skeleton component

BiswaViraj committed:
Merge branch 'next' into nv-5276-ui-overview-tab

BiswaViraj committed:
refactor(subscriber): remove unused Skeleton import from subscriber overview form

BiswaViraj committed:
feat(icons): add CountryFlags component for rendering SVG country flags

BiswaViraj committed:
refactor(subscriber): simplify form submission handling in subscriber overview form

BiswaViraj committed:
feat(subscriber): add timezone and locale selection to subscriber overview form

BiswaViraj committed:
feat(subscriber): implement subscriber deletion with confirmation modal and toast notifications

BiswaViraj committed:
feat(subscriber): reset form with updated subscriber data on successful patch

ChmaraX committed:
Merge branch 'next' into nv-5276-ui-overview-tab

BiswaViraj committed:
refactor(subscriber): change default exports to named exports for SubscriberOverviewForm and useFetchSubscriber

BiswaViraj committed:
Merge branch 'nv-5276-ui-overview-tab' into locale-dropdown

BiswaViraj committed:
Merge branch 'next' into nv-5276-ui-overview-tab

BiswaViraj committed:
Merge branch 'nv-5276-ui-overview-tab' into locale-dropdown

BiswaViraj committed:
Merge branch 'next' into nv-5276-ui-overview-tab

BiswaViraj committed:
Merge branch 'nv-5276-ui-overview-tab' into locale-dropdown

BiswaViraj committed:
fix(subscriber-overview-form): Simplify layout structure by removing unnecessary class

BiswaViraj committed:
fix(subscriber-overview-form): Remove unused imports and simplify subscriber fetching logic

BiswaViraj committed:
fix(subscriber-overview-form): Refactor toast notifications to use dedicated success and error functions

BiswaViraj committed:
Merge branch 'next' into nv-5276-ui-overview-tab

BiswaViraj committed:
Merge branch 'nv-5276-ui-overview-tab' into locale-dropdown

BiswaViraj committed:
fix(locale-select): Adjust layout and enhance locale display in dropdown
fix(subscriber-overview-form): Improve data handling for form submission

BiswaViraj committed:
fix(phone-input): Replace icons with react-icons and enhance styling in country select

BiswaViraj committed:
fix(subscriber-tabs): Simplify subscriber profile display by removing placeholder text

BiswaViraj committed:
fix(subscriber-overview): Enhance form handling and improve layout for subscriber details

BiswaViraj committed:
fix(subscriber-overview-form): Improve form handling and add error messages for better user feedback

BiswaViraj committed:
Merge branch 'next' into nv-5276-ui-overview-tab

BiswaViraj committed:
Merge branch 'nv-5276-ui-overview-tab' into locale-dropdown

BiswaViraj committed:
feat(phone-input): add international prop to PhoneInput component

BiswaViraj committed:
feat(subscribers): add create subscriber functionality with validation

BiswaViraj committed:
fix(create-subscriber): correct typo in variable name for existing subscriber check

BiswaViraj committed:
Merge branch 'next' into nv-5276-ui-overview-tab

BiswaViraj committed:
Merge branch 'nv-5276-ui-overview-tab' into locale-dropdown

BiswaViraj committed:
feat(subscribers): update subscriber creation and preferences handling, improve documentation

BiswaViraj committed:
test(subscribers): add end-to-end test for subscriber creation

BiswaViraj committed:
feat(subscriber-drawer): add onOpenChange prop to handle drawer state changes

BiswaViraj committed:
test(subscribers): enhance end-to-end tests for subscriber creation and error handling

BiswaViraj committed:
test(subscribers): update end-to-end tests to use legacy preferences methods

BiswaViraj committed:
test(subscribers): update end-to-end tests to use legacy preferences methods

BiswaViraj committed:
Merge branch 'next' into nv-5276-ui-overview-tab

BiswaViraj committed:
Merge branch 'nv-5276-ui-overview-tab' into locale-dropdown

BiswaViraj committed:
fix(subscriber-overview-form): update fetching state variable and improve form handling

BiswaViraj committed:
test(subscribers): update end-to-end tests to use legacy preferences methods

BiswaViraj committed:
test(subscribers): update end-to-end tests to use legacy update methods

BiswaViraj committed:
test(subscribers): fix typo in test description for subscriber creation

BiswaViraj committed:
fix(subscribers): change error status code to 409 for existing subscriberId

BiswaViraj committed:
feat(subscribers): add create subscriber form and page with routing

BiswaViraj committed:
feat(subscribers): add CreateSubscriberFormSchema and remove unused subscriber-common-fields component

BiswaViraj committed:
Merge remote-tracking branch 'origin/next' into nv-5181-subscriber-creation-via-ui-UI

BiswaViraj committed:
feat(subscribers): enhance CreateSubscriberForm with environment routing and default values

BiswaViraj committed:
feat(subscribers): add createSubscriber function and useCreateSubscriber hook

BiswaViraj committed:
feat(subscribers): integrate createSubscriber functionality with enhanced form handling and validation

BiswaViraj committed:
fix(subscribers): improve error handling in CreateSubscriberForm on subscriber creation failure

BiswaViraj committed:
feat(subscribers): enforce subscriberId validation and improve form handling in CreateSubscriberForm

BiswaViraj committed:
feat(subscribers): add unsaved changes handling and alert dialog in CreateSubscriberForm and SubscriberOverviewForm

BiswaViraj committed:
refactor(subscribers): remove debug console log from CreateSubscriberForm

BiswaViraj committed:
feat(subscribers): enhance email validation and improve form reset handling in CreateSubscriberForm

BiswaViraj committed:
fix(subscribers): correct typo in error message for subscriber creation failure in CreateSubscriberForm

BiswaViraj committed:
feat(subscribers): add size prop to form controls in SubscriberOverviewForm for improved layout

BiswaViraj committed:
refactor(subscribers): remove unused console log from CreateSubscriberForm

BiswaViraj committed:
Merge branch 'next' into nv-5181-subscriber-creation-via-ui-UI

BiswaViraj committed:
feat(subscriber-overview-form): refactor patchSubscriber logic and improve isDirty check

BiswaViraj committed:
feat(locale-select): implement virtualized rendering for locale options and update flag display

BiswaViraj committed:
feat(dashboard): add search functionality to locale and timezone

BiswaViraj committed:
chore(dependencies): remove unused @tanstack/react-virtual package and update acorn version

BiswaViraj committed:
Merge remote-tracking branch 'origin/next' into locale-optimisation

BiswaViraj committed:
No code changes detected.

BiswaViraj committed:
feat(select): implement open state management for locale and timezone popovers

BiswaViraj committed:
fix(select): update locale and timezone components for improved fallback display

BiswaViraj committed:
feat(select): enhance locale and timezone components with selection indicators

LetItRock commented on apps/dashboard/src/components/subscribers/locale-select.tsx:
```diff
@@ -1,71 +1,111 @@
-  onValueChange: (val: string) => void;
+  onChange: (val: string) => void;
 }) {
+  const [open, setOpen] = useState(false);
```
the list is not scrolling up when the search is performed:
 

https://github.com/user-attachments/assets/3cf41700-14ce-4471-921a-00f5be1041ef

LetItRock replied:
same thing in the timezone

BiswaViraj replied:
It's a bug in the cmdk library😓 
will try some hacks

LetItRock replied:
:/

LetItRock commented on apps/dashboard/src/components/subscribers/locale-select.tsx:
```diff
@@ -1,71 +1,111 @@
+          <div className="flex max-w-full flex-1 items-center gap-1 overflow-hidden">
+            <span className="inline-flex gap-1">
+              {CurrentFlag && <CurrentFlag title={currentCountryCode} className="size-4" />}
+              {value || <span className="text-neutral-400">locale</span>}
```
IDK why but the locale dropdown font size is bigger that the timezone, we should use the same font across the board

LetItRock commented on apps/dashboard/src/components/subscribers/locale-select.tsx:
```diff
@@ -1,71 +1,111 @@
+        <Command>
+          <CommandInput placeholder="Search locale..." />
+          <CommandList>
+            <CommandEmpty>No locale found.</CommandEmpty>
```
this shouldn't be scrollable
![Screenshot 2025-02-07 at 18 44 19](https://github.com/user-attachments/assets/3f1013a3-4496-4282-9e99-d95ac17e4855)

LetItRock replied:
same timezone

LetItRock commented on apps/dashboard/src/components/subscribers/locale-select.tsx:
```diff
@@ -1,71 +1,111 @@
 import TruncatedText from '../truncated-text';
+import { useState } from 'react';
 
 export function LocaleSelect({
```
shouldn't it be possible to clear the locale and timezone? I don't see the option

BiswaViraj replied:
now that you mention it none of our Select fields have clear functionality 😓 🤦🏼

BiswaViraj replied:
Will discuss with Naveen how we will want to introduce a clear functionality

BiswaViraj replied:
Will address this in a different PR as this is something that needs work on the DS level too

BiswaViraj committed:
Merge branch 'next' into locale-optimisation

BiswaViraj committed:
refactor(select): remove ScrollArea component for improved layout

BiswaViraj committed:
fix(select): update text color for improved visibility in locale and timezone selectors

BiswaViraj committed:
feat(select): implement scroll-to-top functionality in locale and timezone selectors

BiswaViraj committed:
Merge branch 'next' into locale-optimisation

BiswaViraj committed:
Merge branch 'next' into locale-optimisation

BiswaViraj committed:
Merge branch 'next' into locale-optimisation

SokratisVidros commented on apps/dashboard/src/components/subscribers/timezone-select.tsx:
```diff
@@ -1,69 +1,102 @@
+              // the setTimeout is used to create a new task
+              // this is to make sure that we don't scroll until the user is done typing
+              // you can tweak the timeout duration ofc
+              scrollId.current = setTimeout(() => {
```
I am curious if [scrollIntoView](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView) helps with this.

BiswaViraj replied:
Had tried that It didnt work

SokratisVidros commented on apps/dashboard/src/components/subscribers/locale-select.tsx:
```diff
@@ -1,71 +1,129 @@
+      onSelect={() => onChange(optionValue)}
+    >
+      <div className="flex w-full items-center gap-2">
+        <div>{CurrentFlag && <CurrentFlag className="size-4" title={countryCode} />}</div>
```
Consider avoid having an empty div.

```suggestion
        { CurrentFlag && <div><CurrentFlag className="size-4" title={countryCode} /></div>}
```

SokratisVidros commented on apps/dashboard/src/components/subscribers/locale-select.tsx:
```diff
@@ -1,71 +1,129 @@
+    <Popover open={open} onOpenChange={setOpen}>
+      <PopoverTrigger asChild>
+        <Button
+          type="button"
```
Is this necessary? What's the default type of the Button?

SokratisVidros commented on apps/dashboard/src/components/subscribers/locale-select.tsx:
```diff
@@ -1,71 +1,129 @@
-              </>
-            )}
+          <div className="flex max-w-full flex-1 items-center gap-1 overflow-hidden">
+            <span className="inline-flex gap-1 text-xs font-normal text-neutral-950">
```
```suggestion
            { (CurrentFlag || value) && <span className="inline-flex gap-1 text-xs font-normal text-neutral-950"> ... }
```

BiswaViraj committed:
Merge branch 'next' into locale-optimisation

BiswaViraj committed:
refactor(subscribers): Improve subscriber form and timezone selection handling

BiswaViraj committed:
fix(hooks): Optimize subscriber patching by setting query data directly

BiswaViraj committed:
fix(phone-input): Disable autocomplete for phone input fields

BiswaViraj committed:
feat(phone-input): Replace fallback icon with phone icon in phone input component
feat(subscriber-overview-form): Add mail icon to leading icon in subscriber overview form

BiswaViraj committed:
fix(subscribers): Update placeholder text in subscriber forms to empty JSON object

BiswaViraj committed:
fix(edit-subscriber-page): Update navigation to include environment slug when unmounting

BiswaViraj committed:
fix(create-subscriber-form): Change form validation mode to 'onBlur'

BiswaViraj committed:
Merge branch 'next' into locale-optimisation

BiswaViraj committed:
fix(subscriber-overview-form): Ensure default values are set to empty strings or null for optional fields

BiswaViraj committed:
fix(subscriber-overview-form): Set email default value to null for better handling of optional fields

BiswaViraj committed:
fix(subscriber-overview-form): Update email handling to use logical OR for default value and set empty string to null on change

BiswaViraj committed:
fix(subscriber-overview-form): Remove debug console logs for cleaner code
